cpp/order/SolidLiquid.cc

```diff
@@ -0,0 +1,101 @@
+// Copyright (c) 2010-2019 The Regents of the University of Michigan
+// This file is from the freud project, released under the BSD 3-Clause License.
+
+#include "SolidLiquid.h"
+
+namespace freud { namespace order {
+
+SolidLiquid::SolidLiquid(unsigned int l, float Q_threshold, unsigned int S_threshold, bool normalize_Q)
+    : m_l(l), m_num_ms(2 * l + 1), m_Q_threshold(Q_threshold), m_S_threshold(S_threshold),
+    m_normalize_Q(normalize_Q), m_steinhardt(l), m_cluster()
+{
+    if (m_Q_threshold < 0.0)
+    {
+        throw std::invalid_argument(
+            "SolidLiquid requires that the dot product cutoff Q_threshold must be non-negative.");
+    }
+}
+
+void SolidLiquid::compute(const freud::locality::NeighborList* nlist,
+        const freud::locality::NeighborQuery* points, freud::locality::QueryArgs qargs)
+{
+    // Make NeighborList from NeighborQuery if not provided
+    if (nlist == NULL)
+    {
+        auto nqiter(points->query(points->getPoints(), points->getNPoints(), qargs));
+        nlist = nqiter->toNeighborList();
+    }
+
+    const unsigned int num_query_points(nlist->getNumQueryPoints());
+
+    // Compute Steinhardt using neighbor list (also gets Ql for normalization)
+    m_steinhardt.compute(nlist, points, qargs);
+    const auto& Qlm = m_steinhardt.getQlm();
+    const auto& Ql = m_steinhardt.getQl();
+
+    // Compute (normalized) dot products for each bond in the neighbor list
+    const float normalizationfactor = float(4 * M_PI / m_num_ms);
+    const unsigned int num_bonds(nlist->getNumBonds());
+    m_Ql_ij.prepare(num_bonds);
+
+    util::forLoopWrapper(0, num_query_points, [=](size_t begin, size_t end) {
+        for (unsigned int i = begin; i != end; ++i)
+        {
+            unsigned int bond(nlist->find_first_index(i));
+            for (; bond < num_bonds && nlist->getNeighbors()(bond, 0) == i; ++bond)
+            {
+                const unsigned int j(nlist->getNeighbors()(bond, 1));
+
+                // Accumulate the dot product over m of Qlmi and Qlmj vectors
+                std::complex<float> bond_Ql_ij = 0;
+                for (unsigned int k = 0; k < m_num_ms; k++)
+                {
+                    bond_Ql_ij += Qlm(i, k) * conj(Qlm(j, k));
+                }
+
+                // Optionally normalize dot products by points' Ql values,
+                // accounting for the normalization of Ql values
+                if (m_normalize_Q)
+                {
+                    bond_Ql_ij *= normalizationfactor / (Ql[i] * Ql[j]);
+                }
+                m_Ql_ij[bond] = bond_Ql_ij.real();
+            }
+        }
+    }, true);
```

Comment:
I'm wondering if we should add a variation on the existing `loopOver*` functions that's specifically for `NeighborLists` and in addition to providing the `NeighborBond` instance, also provides the index in the `NeighborList` of the current bond. The logic would be very similar to what's already in the `NeighborListPerPointIterator`, and it would make these kinds of loops simpler since they're in multiple places (`LocalDescriptors`, `LocalBondProjection`, `AngularSeparation`, etc). Thoughts?

Reply:
Good idea, could be helpful, but not something I want in this PR and not a blocker for 2.0.